Ospi bus support for the stm32H7x devices

FRASTM opened the pull request:
This PR adds the support of the octoSPI bus for the stm32H7 devices
It updates the ospi flash controller driver
It also adds the OSPI instances to the stm32H7 mcus and boards

Not that DMA transfer is not yet in. There is no DMAMUX request for the octospi, only MDMA req 0x16, but the periph is not yet in.

requires the https://github.com/zephyrproject-rtos/zephyr/pull/45487

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32h72x and h73x have octospi peripherals

Add the octospi nodes to the stm32h723/h725
and h730/h733/h735 devices.
Note that RM0468 gives a size of 1kB.
The definition of the octospi clock source for both intances
is given by the DTS node. The selected value for
the alternate clock control is the PLL1_Q.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32h7Ax/h7Bx have octospi peripĥerals

Add the octospi nodes to the stm32h7A3 and
and h7B0/h7B3 devices.
Note that RM0455 gives a size of 1kB.
The definition of the octospi clock source for both intances
is given by the DTS node. The selected value for
the alternate clock control is the PLL1_Q.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32h735g disco target board has octo SPI instance

This commit enables the octo SPI peripheral to the flash-nor
on the target board stm32h735g_disco from STMicroelectronics.
The octospi1 is connected to the NOR- octo-flash through
the OSPI IO Manager.
Note that JESD16 requires 9 dwords for the sdfp table.
The NOR octoflash is MX25LM51245 or compatible.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32h7b3i disco kit board has octo SPI instance

This commit enables the octo SPI peripheral to the flash-nor
on the target board stm32h7b3i_dk from STMicroelectronics.
The octospi1 is connected to the NOR- octo-flash through
the OSPI IO Manager.
Note that JESD16 requires 9 dwords for the sdfp table.
The NOR octoflash is MX25LM51245 or compatible.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: octospi drivers for stm32h7 serie

With the introduction of the OSPI NOR flash controller
the stm32H7 serie requires the HAL MDMA in anycase.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers flash add a test case for stm32h7 octospi

This add a removes the SPI config
for the stm32H735 and stm32H7b3 disco boards
when testing the octoSPI.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

manoj153 commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1160,8 +1160,11 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	/* Initializes the independent peripherals clock */
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
-
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_PLL);
+#else
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
should we refactor to 
```
#if defined(CONFIG_SOC_SERIES_STM32H7X)
#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_PLL
#else
#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_SYSCLK
#endif
```

and then

```
__HAL_RCC_OSPI_CONFIG(OSPI_CLK_SRC);
```

FRASTM replied:
done

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -935,8 +935,8 @@ static void flash_stm32_ospi_isr(const struct device *dev)
 
-/* weak function required for HAL compilation */
-__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
+/* weak function required for HAL compilation */__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
```
```suggestion
#if !defined(CONFIG_SOC_SERIES_STM32H7X)
/* weak function required for HAL compilation */
__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
```
Is slightly easier to read ;)

FRASTM replied:
yes, changed

FRASTM commented:
rebase and fix merge conflicts

FRASTM committed:
include: clock: stm32h7 helper for this serie

For the stm32h7x3 and stm32h7x5 there is a octospi
clock configuration in the D1CCIPR register

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1160,8 +1160,11 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	/* Initializes the independent peripherals clock */
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
-
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_PLL);
+#else
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Ok, It measns that the OSPI clock will comes from the  board DTS, overwriting the property as follows:
```
 &octospi1 {
	/delete-property/ clocks;
	clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00004000>,
		<&rcc STM32_SRC_PLL1_Q OSPI_SEL(1)>;
```
And then the flash_stm32_ospi_init function of the ospi driver will configure the clock with the clock_control_on function.
--> I propose another PR for that change

FRASTM replied:
DNM: waiting for the alternate clock model

FRASTM commented:
rebase
add the OSPI_CLK_SRC to configure the independent peripheral clock (until the  https://github.com/zephyrproject-rtos/zephyr/pull/46703 is completed)
Note the compilation warning  on #define FLASH_SECTOR_SIZE        4096 --> fixed by https://github.com/zephyrproject-rtos/zephyr/pull/45487

FRASTM commented:
tested on the stm32h735g_disco with  west build -p auto -b stm32h735g_disco samples/drivers/spi_flash
```

*** Booting Zephyr OS build zephyr-v3.1.0-589-ge9bd6e5b4bee  ***
JEDEC OSPI-NOR SPI flash testing                                            
==========================  
Test 1: Flash erase       
Flash erase succeeded! 

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!! 
```

FRASTM commented:
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/46842 
Align on the https://github.com/zephyrproject-rtos/zephyr/pull/46703  to define the alternate clock source in the DTS of the soc for each octospi peripheral to be **pll1_q**
The default clock source (0) for the octospi is not the PCLK1 (but _rcc_hclk3_) so it must be defined explicitly.

carlescufi commented:
@manoj153 please revisit

GeorgeCGV reviewed (dismissed):
LGTM

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1569,7 +1576,7 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	/* Initializes the independent peripherals clock */
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
+	__HAL_RCC_OSPI_CONFIG(OSPI_CLK_SRC);
 
```
rebased

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -41,6 +41,12 @@ LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
 /* used as default value for DTS writeoc */
 #define SPI_NOR_WRITEOC_NONE 0xFF
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_PLL
+#else
+#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_SYSCLK
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
```
```suggestion
```

Can be removed as it is not used anymore.

FRASTM replied:
line removed

FRASTM commented:
rebase on 60271524087bb4382c8b4db0f2e17a87e4120065
remove unused OSPI_CLK_SRC definition

GeorgeCGV reviewed (approved):
LGTM

fabiobaltieri commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -224,3 +224,44 @@
+
+	mx25lm51245: ospi-nor-flash@0 {
+		compatible = "st,stm32-ospi-nor";
+		label = "MX25LM51245";
```
Label property are being wiped out all over from the repository, https://github.com/zephyrproject-rtos/zephyr/pull/47881 wiped them from this binding. Can you drop them from here as well?
